File: leetcode/solutions/0733.FloodFill/floodFill.ts
export default function floodFill(
  image: number[][],
  sr: number,
  sc: number,
  newColor: number,
): number[][] {
  if (image[sr][sc] === newColor) {
    return image;
  }
  fill(image, sr, sc, image[sr][sc], newColor);
  return image;
}

const fill = (
  image: number[][],
  n: number,
  m: number,
  color: number,
  newColor: number,
): void => {
  if (image[n] == null || image[n][m] == null || image[n][m] !== color) {
    return;
  }
  image[n][m] = newColor;
  fill(image, n - 1, m, color, newColor);
  fill(image, n + 1, m, color, newColor);
  fill(image, n, m - 1, color, newColor);
  fill(image, n, m + 1, color, newColor);
}

File: leetcode/solutions/0733.FloodFill/floodFill.go
package solution

func floodFill(image [][]int, sr int, sc int, newColor int) [][]int {
	oldColor := image[sr][sc]

	if oldColor == newColor {
		return image
	}

	fill(image, sr, sc, newColor, oldColor)

	return image
}

func fill(image [][]int, sr, sc, newColor, oldColor int) {
	if sr < 0 || sc < 0 || sr >= len(image) || sc >= len(image[sr]) || image[sr][sc] != oldColor {
		return
	}

	image[sr][sc] = newColor

	fill(image, sr+1, sc, newColor, oldColor)
	fill(image, sr-1, sc, newColor, oldColor)
	fill(image, sr, sc+1, newColor, oldColor)
	fill(image, sr, sc-1, newColor, oldColor)
}
